p2p/src/utils/mux.rs

```diff
@@ -0,0 +1,358 @@
+//! This utility wraps a [Sender] and [Receiver], providing lightweight sub-channels keyed by the
+//! [Channel] type.
+//!
+//! Usage:
+//! - Call [Muxer::new] to create the multiplexer.
+//! - Call [Muxer::register] to obtain a `(SubSender, SubReceiver)` pair for that subchannel.
+//! - Drive [Muxer::run] in a background task to demux incoming messages into per-subchannel queues.
+use crate::{Channel, Message, Receiver, Recipients, Sender};
+use bytes::{BufMut, Bytes, BytesMut};
+use commonware_codec::{varint::UInt, EncodeSize, ReadExt, Write};
+use commonware_cryptography::PublicKey;
+use commonware_runtime::{Handle, Spawner};
+use futures::{channel::mpsc, StreamExt};
+use std::{
+    collections::HashMap,
+    sync::{Arc, Mutex},
+};
+use thiserror::Error;
+
+/// An asynchronous map of [Channel]s to [mpsc::Sender]s.
+type Routes<P> = Arc<Mutex<HashMap<Channel, mpsc::Sender<Message<P>>>>>;
+
+/// Errors that can occur when interacting with a [Muxer].
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("recv failed")]
+    RecvFailed,
+}
+
+/// A multiplexer of p2p channels into subchannels.
+pub struct Muxer<S: Sender, R: Receiver> {
+    sender: S,
+    receiver: R,
+    #[allow(clippy::type_complexity)]
+    routes: Routes<R::PublicKey>,
+    mailbox_size: usize,
+}
+
+impl<S: Sender, R: Receiver> Muxer<S, R> {
+    /// Create a multiplexed wrapper around a [Sender] and [Receiver] pair.
+    pub fn new(sender: S, receiver: R, mailbox_size: usize) -> Self {
+        Self {
+            sender,
+            receiver,
+            routes: Arc::new(Mutex::new(HashMap::new())),
+            mailbox_size,
+        }
+    }
+
+    /// Open a `subchannel`. Returns a `(SubSender, SubReceiver)` pair that can be used to send and
+    /// receive messages for that subchannel.
+    ///
+    /// The caller must be driving [Muxer::run], or have started it using [Muxer::start], for the
+    /// receiver to yield messages.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the subchannel is already registered.
+    pub fn register(&self, subchannel: Channel) -> (SubSender<S>, SubReceiver<R::PublicKey>) {
+        // Create a new channel to forward messages to the subchannel.
+        let (tx, rx) = mpsc::channel(self.mailbox_size);
+
+        // Insert the subchannel into the routes map, panicking if it already exists.
+        if self.routes.lock().unwrap().insert(subchannel, tx).is_some() {
+            panic!("duplicate subchannel registration: {subchannel}");
+        }
+
+        // Return the subchannel sender and receiver.
+        (
+            SubSender {
+                subchannel,
+                inner: self.sender.clone(),
+            },
+            SubReceiver {
+                subchannel,
+                receiver: rx,
+                routes: Arc::clone(&self.routes),
```

Comment:
why not `self.routes.clone()`?

Reply:
Ah, didn't realize it was just the pointer. I suppose that clippy recommendation makes sense.